win32/src/job_object_helper.c

```diff
@@ -174,10 +174,148 @@ IMPLEMENT_MOCKABLE_FUNCTION(, int, job_object_helper_limit_cpu, THANDLE(JOB_OBJE
+        if (job_object_helper == NULL)
+        {
+            /*Codes_SRS_JOB_OBJECT_HELPER_19_009: [ If there are any failures, job_object_helper_set_job_limits_to_current_process shall fail and return NULL. ]*/
+            LogError("failure in THANDLE_MALLOC(JOB_OBJECT_HELPER)(job_object_helper_dispose=%p)", job_object_helper_dispose);
+            result = NULL;
+        }
+        else
+        {
+            /*Codes_SRS_JOB_OBJECT_HELPER_19_002: [ job_object_helper_set_job_limits_to_current_process shall call CreateJobObjectA passing job_name for lpName and NULL for lpJobAttributes. ] */
+            job_object_helper->job_object = CreateJobObjectA(NULL, job_name);
+            if (job_object_helper->job_object == NULL)
+            {
+                /*Codes_SRS_JOB_OBJECT_HELPER_19_009: [ If there are any failures, job_object_helper_set_job_limits_to_current_process shall fail and return a non-zero value. ] */
+                LogLastError("failure in CreateJobObjectA(lpJobAttributes=NULL, job_name=%s)", MU_P_OR_NULL(job_name));
+                result = NULL;
+            }
+            else
+            {
+                bool failed = false;
+                /*Codes_SRS_JOB_OBJECT_HELPER_19_003: [ If percent_cpu is not 0 then job_object_helper_set_job_limits_to_current_process shall call SetInformationJobObject passing JobObjectCpuRateControlInformation and a JOBOBJECT_CPU_RATE_CONTROL_INFORMATION object with JOB_OBJECT_CPU_RATE_CONTROL_ENABLE and JOB_OBJECT_CPU_RATE_CONTROL_HARD_CAP set, and CpuRate set to percent_cpu times 100. ]*/
+                if (percent_cpu != 0)
+                {
+                    JOBOBJECT_CPU_RATE_CONTROL_INFORMATION cpu_rate_control_information = { 0 };
+                    cpu_rate_control_information.ControlFlags = JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | JOB_OBJECT_CPU_RATE_CONTROL_HARD_CAP;
+                    cpu_rate_control_information.CpuRate = percent_cpu * 100;
+                    if (!SetInformationJobObject(job_object_helper->job_object, JobObjectCpuRateControlInformation, &cpu_rate_control_information, sizeof(cpu_rate_control_information)))
+                    {
+                        /*Codes_SRS_JOB_OBJECT_HELPER_19_009: [ If there are any failures, job_object_helper_set_job_limits_to_current_process shall fail and return a non-zero value. ] */
+                        LogLastError("failure in SetInformationJobObject(job_object=%p, JobObjectCpuRateControlInformation, &cpu_rate_control_information=%p, sizeof(cpu_rate_control_information)=%zu)",
+                            job_object_helper->job_object, &cpu_rate_control_information, sizeof(cpu_rate_control_information));
+                        result = NULL;
+                        failed = true;
+                    }
+                }
+
+                if (failed)
+                {
+                    LogError("failure in SetInformationJobObject(job_object_helper->job_object=%p)", job_object_helper->job_object);
```

Comment:
>LogError("failure in SetInformationJobObject(job_object_helper->job_object=%p)", job_object_helper->job_object); [](http://example.com/codeflow?start=20&length=112)

logging is redundant, can just have a comment here (like "error logged above" or something)